xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,607 @@
+        """
+        if len(self.results) != 1:
+            raise ValueError(
+                "'result' property of IOperation class is only available "
+                "for IOperations with exactly one result.")
+        return self.results[0]
+
+    @property
+    def region(self) -> IRegion:
+        """
+        Get the region of a of an IOperation with a single region.
+        Returns an exception if the operation does not have exactly one region.
+        """
+        if len(self.regions) != 1:
+            raise ValueError(
+                "'region' property of IOperation class is only available "
+                "for IOperations with exactly one region.")
+        return self.regions[0]
+
+    @property
+    def result_types(self) -> list[Attribute]:
+        return [result.typ for result in self.results]
+
+    def get_mutable_copy(
+            self,
+            value_mapping: dict[ISSAValue, SSAValue] | None = None,
+            block_mapping: dict[IBlock, Block] | None = None) -> Operation:
+        """
+        Returns a mutable operation that is a copy of this immutable operation.
+        The value_mapping and block_mapping are used to map already known correspondings
+        of immutable values to mutable values and immutable blocks to mutable blocks.
+        """
+        if value_mapping is None:
+            value_mapping = {}
+        if block_mapping is None:
+            block_mapping = {}
+
+        mutable_operands: list[SSAValue] = []
+        for operand in self.operands:
+            if operand in value_mapping:
+                mutable_operands.append(value_mapping[operand])
+            else:
+                print(f"ERROR: op {self.name} uses SSAValue before definition")
+                # Continuing to enable printing the IR including missing
+                # operands for investigation
+                mutable_operands.append(
+                    OpResult(
+                        operand.typ,
+                        None,  # type: ignore
+                        0))
+
+        mutable_successors: list[Block] = []
+        for successor in self.successors:
+            if successor in block_mapping:
+                mutable_successors.append(block_mapping[successor])
+            else:
+                raise Exception("Block used before definition")
+
+        mutable_regions: list[Region] = []
+        for region in self.regions:
+            mutable_regions.append(
+                region.get_mutable_copy(value_mapping=value_mapping,
+                                        block_mapping=block_mapping))
+
+        new_op: Operation = self.op_type.create(
+            operands=mutable_operands,
+            result_types=[result.typ for result in self.results],
+            attributes=self.attributes.copy(),
+            successors=mutable_successors,
+            regions=mutable_regions)
+
+        for idx, result in enumerate(self.results):
```

Comment:
Maybe add a comment that this will have an effect when this is called as part of a different conversion. I was personally a bit confused why this is required at first.

Reply:
:+1: